crates/ruff_graph/src/collector.rs

```diff
@@ -0,0 +1,120 @@
+    pub(crate) fn collect(mut self, module: &ModModule) -> Vec<CollectedImport> {
+        walk_body(&mut self, &module.body);
+        self.imports
+    }
+}
+
+impl<'ast> SourceOrderVisitor<'ast> for Collector<'_> {
+    fn visit_stmt(&mut self, stmt: &'ast Stmt) {
+        match stmt {
+            Stmt::ImportFrom(ast::StmtImportFrom {
+                names,
+                module,
+                level,
+                range: _,
+            }) => {
+                let module = module.as_deref();
+                let level = *level;
+                for alias in names {
+                    if level == 0 {
+                        let mut components = vec![];
+
+                        // Add the module path.
+                        if let Some(module) = module {
+                            components.extend(module.split('.'));
+                        }
+
+                        // Add the alias name.
+                        components.push(alias.name.as_str());
+
+                        if let Some(module_name) = ModuleName::from_components(components) {
+                            self.imports.push(CollectedImport::ImportFrom(module_name));
+                        }
+                    } else if let Some(module_path) = self.module_path {
+                        let mut components = vec![];
+
+                        // Start with the containing module.
+                        components.extend(module_path.iter().map(String::as_str));
+
+                        // Remove segments based on the number of dots.
+                        for _ in 0..level {
+                            if components.is_empty() {
+                                return;
+                            }
+                            components.pop();
+                        }
+
+                        // Add the module name.
+                        if let Some(module) = module {
+                            components.extend(module.split('.'));
+                        }
+
+                        // Add the alias name.
+                        components.push(alias.name.as_str());
+
+                        if let Some(module_name) = ModuleName::from_components(components) {
+                            self.imports.push(CollectedImport::ImportFrom(module_name));
+                        }
+                    }
+                }
+            }
+            Stmt::Import(ast::StmtImport { names, range: _ }) => {
+                for alias in names {
+                    if let Some(module_name) = ModuleName::new(alias.name.as_str()) {
+                        self.imports.push(CollectedImport::Import(module_name));
+                    }
+                }
+            }
+            _ => {
+                walk_stmt(self, stmt);
+            }
+        }
+    }
+
+    fn visit_expr(&mut self, expr: &'ast Expr) {
+        if self.string_imports {
+            if let Expr::StringLiteral(ast::ExprStringLiteral { value, range: _ }) = expr {
```

Comment:
This is more aggressive than I realized -- we don't even look for calls to `__import__` or `importlib.import_module`, we just look for any string that might be a module path.

It seems like this could easily be fooled by cases where a module constant is defined in one file (like a Django settings file) but the string is imported and then used in a dynamic import in a different module.

But if it's good enough to match existing tools on the codebases we care about, great!